### PR TITLE
Sales eng add fair value

### PR DESCRIPTION
**Pull request by alexgraham-da**
add the ability for the provider/customer of the `Clearing.Market.Service` to request `FairValue` contracts for listings

A `FairValue` is calculated by:
1. The price of the most recent trade before the specified 'upTo'
2. If there are no trades, the most recent 'FairValue'
3. If there are no 'FairValue', request manual calculation

### CODE REVIEW

File: daml/Marketplace/Clearing/Market/Model.daml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+template FairValueCalculationRequest
+  with
+    operator : Party
+    provider : Party
+    customer : Party
+    optListingIds : Optional [Text]
+    currency : Id
+    upTo : Time
+  where
+    signatory provider
```
Comment (brian-weir) [inline daml/Marketplace/Clearing/Market/Model.daml:15]: Should the `operator` also be a signature ?

**Review by brian-weir — APPROVED** [review]

File: daml/Marketplace/Clearing/Market/Model.daml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    controller provider can
+      FairValueCalculationRequest_Ack : ()
+        do return ()
+
+template ManualFairValueCalculation
+  with
+    provider : Party
```
Comment (brian-weir) [inline daml/Marketplace/Clearing/Market/Model.daml:22]: Should these templates have the `operator : Party` as part of them ?

File: ui2/src/pages/listing/Listing.tsx [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+  const listing = listings.find(a => a.contractId === cid);
+  const service = services.find(s => s.payload.customer === party);
+
+  const { contracts: fairValueContracts, loading: fairValuesLoading } = useStreamQueries(FairValue);
+  const { contracts: clearedMarketServices } = useStreamQueries(ClearedMarketService);
+  const clearedMarketService = clearedMarketServices.find(
+    cms => cms.payload.customer === listing?.payload.customer
+  );
+  console.log(clearedMarketService);
```
Comment (Sarahb-da) [inline ui2/src/pages/listing/Listing.tsx:83]: extra log

File: ui2/src/pages/listing/Listing.tsx [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+            )}
+            <Button floated="left" className="ghost" onClick={() => requestDisableDelisting()}>
+              Disable
+            </Button>
+          </Button.Group>
+        </div>
+      </Tile>
+      {!!listing && (
+        <Tile header={<h4>Details</h4>}>
+          <StripedTable
```
Comment (Sarahb-da) [inline ui2/src/pages/listing/Listing.tsx:126]: u could add `loading` to line 73 and use it here to give this table a loading state

File: ui2/src/pages/listing/Listings.tsx [hunk 5]
```diff
@@ -18,49 +25,88 @@ export const ListingsTable: React.FC<Props> = ({ services, listings }) => {
...
+          'Listing ID',
+          'Calendar ID',
+          'Traded Asset',
+          'Traded Asset Precision',
+          'Quoted Asset',
+          'Quoted Asset Precision',
+          'Fair Value',
+          'Action',
+        ]}
+        rowsClickable
```
Comment (Sarahb-da) [inline ui2/src/pages/listing/Listings.tsx:70]: Sadly I dont think this will work - the row being clickable overlaps with the button in the row. Making the click functionality for the button possibly not work.. its pretty poor design on my part... in order to have clickable rows and buttons there needs to be some kind of click listener added. So the DOM knows if the user clicked the row, or the button in the row.

Comment (Sarahb-da) [reply ui2/src/pages/listing/Listings.tsx:70]: basically `rowsClickable` only works for tables with no buttons in them

File: ui2/src/pages/listing/ManualCalculationRequests.tsx [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+            <Button.Group floated="right">
+              <ModalFormErrorHandled onSubmit={() => applyCalculation(fv)} title="Calculate">
+                <Form.Input
+                  label="Amount"
+                  placeholder="Select..."
+                  required
+                  number
+                  min={1}
+                  value={valueAmount}
+                  onChange={(_, change) => setValueAmount(change.value as string)}
```
Comment (Sarahb-da) [inline ui2/src/pages/listing/ManualCalculationRequests.tsx:50]: does the `rowsClickable` prop mess up the `input`? curious

Comment (Sarahb-da) [reply ui2/src/pages/listing/ManualCalculationRequests.tsx:50]: u actually dont need it anyway bc the row has no `onClick` function

**Review by Sarahb-da — APPROVED** [review]
just a few minor comments but looks great

File: ui2/src/pages/listing/ManualCalculationRequests.tsx (hunk 6 above)
Comment (alexgraham-da) [reply ui2/src/pages/listing/ManualCalculationRequests.tsx:50]: the `Input` is in the `ModalFormErrorHandled` so it only pops up when the modal does. row doesn't need to be clickable though, so I'll remove it

**Review by alexmatson-da — APPROVED** [review]

File: daml/Marketplace/Clearing/Market/Model.daml (hunk 1 above)
Comment (alexgraham-da) [reply daml/Marketplace/Clearing/Market/Model.daml:15]: I can add this. I'm curious though, I've noticed on multiple models I've worked on that the `operator` is a `signatory` on the majority of contracts, even ones the `operator` does not need to perform any action on. What is the general purpose of this? Is it to provide visibility to the ledger administrator (and wouldn't this be a possible privacy concern?). Or maybe to disallow a party from archiving contracts without using a `choice`?

File: ui2/src/pages/listing/Listings.tsx (hunk 5 above)
Comment (alexgraham-da) [reply ui2/src/pages/listing/Listings.tsx:70]: ah OK - I can move that button to the listing's page and look into adding a listener like that (in another PR).

File: daml/Marketplace/Clearing/Market/Model.daml [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+import DA.Finance.Types
+import DA.Next.Set (Set)
+
+template FairValueCalculationRequest
+  with
+    operator : Party
+    provider : Party
+    customer : Party
+    optListingIds : Optional [Text]
```
Comment (lucianojoublanc-da) [inline daml/Marketplace/Clearing/Market/Model.daml:11]: Probably worth `newtype`ing this. How are we currently identifying listings?

File: daml/Marketplace/Clearing/Market/Model.daml [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    provider : Party
+    customer : Party
+    listingId : Text
+    currency : Id
+    upTo : Time
+    observers : Set Party
+  where
+    signatory provider
+
+    controller customer can
```
Comment (lucianojoublanc-da) [inline daml/Marketplace/Clearing/Market/Model.daml:31]: Doesn't make sense to me from a business point of view that the customer is valuing their own positions... Maybe related to @brian-weir 's comment above.

File: daml/Marketplace/Clearing/Market/Service.daml [hunk 9]
```diff
@@ -16,8 +20,18 @@ template Service
     key (operator, provider, customer) : (Party, Party, Party)
     maintainer key._1
 
-    controller customer can
+    nonconsuming choice RequestFairValues : ContractId FairValueCalculationRequest
```
Comment (lucianojoublanc-da) [inline daml/Marketplace/Clearing/Market/Service.daml:23]:
```suggestion
    nonconsuming choice RequestFairValue : ContractId FairValueCalculationRequest
```
Is this a typo?

File: triggers/daml/ClearingTrigger.daml [hunk 10]
```diff
@@ -146,3 +162,52 @@ collectDeposits restFunds (used, (nextDeposit::remaining))
...
+--    2. If there are no trades, the most recent 'FairValue'
+--    3. If there are no 'FairValue', request manual calculation
+--
+-- TODO: Allow user to specify a 'from' Time interval before which trades and
+-- fairvalues will be ignored.
+handleFairValueCalculation : Party -> Id -> Time -> ContractPair Listing.T -> TriggerA ClearingState ()
+handleFairValueCalculation party currency upTo (marketPairCid,l@(Listing.Listing{..})) = do
+  debug ("calculating FVR: " <> l.listingId)
+
+  fairValues <- sortOn ((.upTo) . snd)
```
Comment (lucianojoublanc-da) [inline triggers/daml/ClearingTrigger.daml:193]:
```suggestion
  fairValues <- sortOn (upTo . snd)
```

File: triggers/daml/ClearingTrigger.daml [hunk 11]
```diff
@@ -146,3 +162,52 @@ collectDeposits restFunds (used, (nextDeposit::remaining))
...
+        (Some ids) -> filterField (.listingId) (`Set.member` ids) allListings
+        None       -> allListings
+
+  forA_ listings (handleFairValueCalculation party currency upTo)
+  emitExerciseCmd requestCid Clearing.FairValueCalculationRequest_Ack
+  return ()
+
+-- |Calculates a FairValue for a single derivative.
+-- Order of precedence:
+--    1. The price of the most recent trade before the specified 'upTo'
```
Comment (lucianojoublanc-da) [inline triggers/daml/ClearingTrigger.daml:183]: This doesn't sound right. Shouldn't you be taking price _after_ the specified date?

File: triggers/daml/ClearingTrigger.daml [hunk 12]
```diff
@@ -146,3 +162,52 @@ collectDeposits restFunds (used, (nextDeposit::remaining))
...
+  case getBestPrice relevantTrades fairValues of
+    (Some bp) -> dedupCreate $ Clearing.FairValue with provider = party, price = bp; ..
+    None      -> dedupCreate $ Clearing.ManualFairValueCalculation with provider = party,  ..
+
+  return ()
+
+  where
+    getBestPrice : [ContractPair Clearing.ClearedTradeSide] -> [ContractPair Clearing.FairValue] -> Optional Decimal
+    getBestPrice ts fvs
+      | not (null ts)  = Some . (.price) . (.execution) . snd . head $ ts
```
Comment (lucianojoublanc-da) [inline triggers/daml/ClearingTrigger.daml:211]:
```suggestion
      | not (null ts)  = Some . price . execution . snd . head $ ts
```
Does that not work?

**Review by lucianojoublanc-da — COMMENTED** [review]
I'd like to understand better in what context this is used. Because this is quite a sensitive thing - getting this wrong can lead to large realized losses. It's particularly problematic when you have thinly traded securities, where the price isn't representative at all, so I'm a bit worried about this logic of looking at the last price, or similar ... :grimacing:

File: daml/Marketplace/Clearing/Market/Model.daml (hunk 8 above)
Comment (alexgraham-da) [reply daml/Marketplace/Clearing/Market/Model.daml:32]: The customer here is the listing agent, not a clearinghouse member, this is the same `Service` that the listing agent uses to request the clearinghouse cover their market.  Does it make more sense to you that way? It used to be the exchange handling any manual valuations, and since that is now split out into multiple services, the listing part of it made the most sense to me.

In the end I think this will fit best somewhere in the listing service directly, so any customer can get fair values of instruments (regardless of if they are cleared or not).

File: daml/Marketplace/Clearing/Market/Model.daml (hunk 7 above)
Comment (alexgraham-da) [reply daml/Marketplace/Clearing/Market/Model.daml:11]: Elsewhere they've been `listingId: Text`, but I think using `Id` might be better so you can `key` the `Listing` with the `Id`. Currently if you're trying to do a `fetchByKey` of a listing, you wouldn't know who the provider/customer is to do the lookup.

File: triggers/daml/ClearingTrigger.daml (hunk 12 above)
Comment (alexgraham-da) [reply triggers/daml/ClearingTrigger.daml:211]: It does not...`(.price)` is a field accessor, I believe it's the only way of doing `contract.field` as a single function, at least for a `template`.

File: daml/Marketplace/Clearing/Market/Service.daml (hunk 9 above)
Comment (alexgraham-da) [reply daml/Marketplace/Clearing/Market/Service.daml:23]: It was not since you are able to use this choice to request `FairValue`s for multiple listings. Now that I'm looking at it, it might make some sense to have multiple choices - `RequestFairValue` for one listing, `RequestFairValues` for multiple listings, and `RequestAllFairValues` for all listings instead of having the `Optional` that doesn't really suggest the actual functionality (`None` meaning all listings)...

File: triggers/daml/ClearingTrigger.daml (hunk 11 above)
Comment (alexgraham-da) [reply triggers/daml/ClearingTrigger.daml:183]: `upTo` was how the requirement was given, the idea being "I would like the latest fair value, ignoring anything after 5pm". It might make sense to also have an optional `after : Time` to allow for a range.

File: triggers/daml/ClearingTrigger.daml (hunk 10 above)
Comment (alexgraham-da) [reply triggers/daml/ClearingTrigger.daml:193]: See https://github.com/digital-asset/da-marketplace/pull/300#discussion_r625891796

File: triggers/daml/ClearingTrigger.daml (hunk 12 above)
Comment (alexgraham-da) [reply triggers/daml/ClearingTrigger.daml:211]: Actually, I think you _can_ use the accessors as functions without the `.`, but since we aren't bothering making field names unique, it won't infer which function to use...so since multiple templates in the `Clearing` namespace have a `price` field, you get an ambiguous occurrence error.

**Comment by alexgraham-da** [comment]
> I'd like to understand better in what context this is used. Because this is quite a sensitive thing - getting this wrong can lead to large realized losses. It's particularly problematic when you have thinly traded securities, where the price isn't representative at all, so I'm a bit worried about this logic of looking at the last price, or similar ... 😬

@lucianojoublanc-da - this is all used as a way for the clearinghouse (currently only Baymarkets) to have a way of calculating mark to markets and margins. The algorithm for determining the fair values was a simplistic one to use as a POC and can be changed however is necessary. I think a good way to handle this would be to allow the customer to choose from different methods of calculating fair values, maybe even in a composable way similar to the contingent claims library.

File: daml/Marketplace/Clearing/Market/Model.daml (hunk 7 above)
Comment (alexgraham-da) [reply daml/Marketplace/Clearing/Market/Model.daml:11]: I am addressing this is a separate PR

File: daml/Marketplace/Clearing/Market/Service.daml (hunk 9 above)
Comment (alexgraham-da) [reply daml/Marketplace/Clearing/Market/Service.daml:33]: I split the choice as mentioned above.

File: daml/Marketplace/Clearing/Market/Model.daml (hunk 8 above)
Comment (lucianojoublanc-da) [reply daml/Marketplace/Clearing/Market/Model.daml:32]: I would avoid using 'customer' as it can mean many things. Typically it would not be a 3rd party with no affiliations. Does either 'issuer' or 'exchange' fit the description?

File: triggers/daml/ClearingTrigger.daml [hunk 13]
```diff
@@ -146,3 +162,52 @@ collectDeposits restFunds (used, (nextDeposit::remaining))
...
+--    3. If there are no 'FairValue', request manual calculation
+--
+-- TODO: Allow user to specify a 'from' Time interval before which trades and
+-- fairvalues will be ignored.
+handleFairValueCalculation : Party -> Id -> Time -> ContractPair Listing.T -> TriggerA ClearingState ()
+handleFairValueCalculation party currency upTo (marketPairCid,l@(Listing.Listing{..})) = do
+  debug ("calculating FVR: " <> l.listingId)
+
+  fairValues <- sortOn ((.upTo) . snd)
+            <$> filterQuery @Clearing.FairValue (.listingId) (== l.listingId)
```
Comment (lucianojoublanc-da) [inline triggers/daml/ClearingTrigger.daml:197]: Btw not sure this `filterQuery` is helping at all; think that in-lining this with a lambda would be much more legible.

File: triggers/daml/ClearingTrigger.daml (hunk 11 above)
Comment (lucianojoublanc-da) [reply triggers/daml/ClearingTrigger.daml:186]: So what if
- the trade was at 9 am this morning? Will that be ignored because it's after 5 pm yesterday?
- the last trade was three days ago? (This happens in real life).

> upTo was how the requirement was given,

I'd like to understand if this was an internal requirement / Exberry / a client. Maybe we can chat off-line.

File: triggers/daml/ClearingTrigger.daml (hunk 13 above)
Comment (alexgraham-da) [reply triggers/daml/ClearingTrigger.daml:197]: I personally like the way it looks since if you read through the lambdas it reads as an expression (`Template.fieldName == variable)` but if you're strongly against it I can change. I find the lambdas pattern matching the tuples all over these triggers ugly, but it's readable in that it is exactly how it looks, so that's fine.

File: triggers/daml/ClearingTrigger.daml (hunk 11 above)
Comment (alexgraham-da) [reply triggers/daml/ClearingTrigger.daml:186]: This was an Eric requirement, but we can chat about it. For the first one, yes, if you specified 5pm yesterday then the trade was at 9am this morning will be ignored by design, since you're interested in doing calculations based on what happened yesterday. For the second, also yes based on this particular algorithm (which is overly simple and can be changed) since that is the most recent traded value.

File: daml/Marketplace/Clearing/Market/Model.daml (hunk 8 above)
Comment (alexgraham-da) [reply daml/Marketplace/Clearing/Market/Model.daml:32]: I was going with `customer` to match the service but I would prefer `exchange` anyway.

Comment (brian-weir) [reply daml/Marketplace/Clearing/Market/Model.daml:32]: Almost all our services have `operator, provider, customer` parties as part of their template and also the template's signature. It makes it clear which party is the producer and which is the consumer of said services. Plus, once we get generics (or what they are calling 'interfaces') back into Daml, it should (in theory...) be easy to extract these details out of all our services (along with the offer/request templates) if we're consistent in the code base.

I do understand your point that `exchange` provides more context to the expected party... would something like the below which addresses both concerns help ? 

> let exchange = customer
>
> controller exchange can

File: daml/Marketplace/Clearing/Market/Model.daml (hunk 1 above)
Comment (brian-weir) [reply daml/Marketplace/Clearing/Market/Model.daml]: Good question - this main use of this pattern is to enforce that contracts are created only via service `choice`s which require the signature from the `operator` of that Marketplace. In other words, a `provider` (or a `customer` or even an `operator`...) can't simply create contracts outside of the "approved" service context, they must follow the explicit workflow provided by a service.

Regarding the privacy concern of this pattern, it's correct that an `operator` will have visibility of all the contracts they are a signatory to. Whether this is an issue or not likely depends on how you view the the `operator` which runs such a Marketplace. Should this be viewed as `admin` access (akin to a database super user access) which can view and monitor an entire Marketplace (makes it easier to audit, regulatory reporting, etc.) ? Or, should templates only have parties which are strictly required be allowed ? I think there's a trade off to be made here and currently we're enforcing having the correct signatures over privacy.

@GeorgSchneider , do you have view on this ?

Comment (GeorgSchneider) [reply daml/Marketplace/Clearing/Market/Model.daml]: yep, having the operator on there is merely to be able to restrict "valid" contracts in an operator-centric marketplace. That is, a given marketplace (operator) will only accept contracts that are signed by the operator, such that different parties cannot act "outside" of the provided service framework. In the decentralized operating model, the services would be "self-signed" with the providers as the respective operators. Re privacy, it is by design that an operator-centric marketplace would give full visibility for the operator. The decentralized mode is one that would retain certain privacy properties.

In summary I'd suggest to add the operator here as the signatory, assuming that the service this originates from does have the operator signature on it so it can be done easily.

Comment (alexgraham-da) [reply daml/Marketplace/Clearing/Market/Model.daml]: This makes a lot of sense, thanks all for the explanations. I actually noticed when adding the `operator` that I was required then to create these contracts in the trigger via the `Service`, so that gave me a nice practical example of why this pattern is useful 😃 .
